[Delta-474][Delta-481] - Use InitContext#getRestoredCheckpointId to get checkpointID needed by Delta writer. Flink 1.15.3 update.

kristoffSC opened the pull request:
Solve https://github.com/delta-io/connectors/issues/481 and https://github.com/delta-io/connectors/issues/474

Problem:
Currently `DeltaWritter` uses checkpoint ID stored in `DeltaWriterBucketState`.
This might not work well for cluster up scaling since this could potentially lead to that the task start with checkpointId = 0 although it is recovering from a checkpoint. Upscale/downscale Flink job requires job restart which will trigger recovery from a last checkpoint. 

In this case, if there will be a new DeltaWritter created from upscale event and there will be no state to recover from for this task, in old implementation, this new DeltaWritter will have checkpointid = 0 assigned which may not be a correct value.

This has no impact on data correctness but it might produce confusing application log data since Committable objects created by new writers will use wrong checkpointId value that is used in logs.

Instead keeping `checkpointId` in bucket state, the `checkpointId` should be recovered/acquired from Flink's `InitContext#getRestoredCheckpointId` since this will provide current global checkpointId for all tasks, not relaying on individual task's state that might be empty for new tasks created from upscale event.

Solution:
1. Migrate Flink connector to Flink version 1.15.3 - un-comment tests that were waiting for Flink 1.15.3

2. Use InitContext#getRestoredCheckpointId to get checkpointID needed by Delta writer instead recover checkpoint Id from state.

fapaul commented on flink/src/main/java/io/delta/flink/sink/internal/writer/DeltaWriterBucket.java:
```diff
@@ -232,15 +232,10 @@ List<DeltaCommittable> prepareCommit(boolean flush,
-            appId,
-            checkpointId
+    DeltaWriterBucketState snapshotState(String appId) {
+        return new DeltaWriterBucketState(bucketId, bucketPath, appId
```
Nit: 
```suggestion
        return new DeltaWriterBucketState(bucketId, bucketPath, appId);
```

kristoffSC replied:
Done

fapaul commented on flink/src/main/java/io/delta/flink/sink/internal/writer/DeltaWriter.java:
```diff
@@ -238,14 +237,14 @@ public List<DeltaWriterBucketState> snapshotState() {
         }
 
         if (states.isEmpty()) {
             // we still need to snapshot transactional ids (appId and checkpointId) even though
```
```suggestion
            // we still need to snapshot transactional ids (appId) even though
```

kristoffSC replied:
Done

fapaul commented on flink/src/test/java/io/delta/flink/sink/internal/writer/DeltaWriterBucketStateSerializerTest.java:
```diff
@@ -0,0 +1,58 @@
+    @Test
+    public void testDeserializeV1State() throws IOException {
+        java.nio.file.Path v1StatePath =
+            Paths.get("src/test/resources/state/bucket-writer/DeltaWriterBucketStateV1.ser");
```
Nit: Since probably in a few years noone remembers how to generate that file it would be good to have a quick readme or even a better code where we can easily regenerate it.

kristoffSC replied:
But for "code" we would need to have old version of DeltaWriterBucketState state, the one that has checkpointId field.

kristoffSC replied:
To generate the DeltaWriterBucketStateV1.ser file (V1 is an old version) I used old/previous version of 'DeltaWriterBucketStateSerializer', I just used this class to serialize DeltaWriterBucketState and I use wrote created byte[] to this file.

And we don't have old DeltaWriterBucketStateSerializer anymore :)

fapaul reviewed (commented):
Overall looks good, but I left some nitty comments.

In general, interpreting the checkpointId can still be very hard since the checkpointId we forward from the writer might not be fully accurate. It is possible that `prepareCommit` is called without a following checkpoint i.e. if the pipeline finishes or for batch executions.

I am okay with keeping it as a general monitoring tool, but we should be aware of its limitations.

tdas commented:
@kristoffSC can you explain in the description why this change in how to get the checkpoint id. I dont get why this new approach is better.

scottsand-db commented:
Looking good! Please implement Fabian's feedback and update the description as TD asked for, and then will LGTM!

kristoffSC commented:
I've implemented requested changes.

@tdas I've change the description, please let me know what do you think.
@fapaul I've added extra comment to DeltaWriter::nextCheckpointId field describing this value and its limitations.
